### DISCUSSION TITLE
Use transforms: ["middle-out"]

### DISCUSSION THREAD

**Discussion by raymatos** [cline/cline#743] [Feature Requests]
When using open router we have the ability to use middle-out to compress the context for those lengthly code files.

https://openrouter.ai/docs/transforms

**Comment by deepfold9118** [comment]
But how does one use this with Cline? I don't know of a way to add that code to a cline task?

**Comment by lloydchang** [comment]
relates to https://discord.com/channels/1275535550845292637/1275535550845292640/1311469773208420492

if anyone wants to experiment with https://openrouter.ai/docs/transforms 
```TypeScript
transforms: ["middle-out"],
// Compress prompts > context size.
// This is the default for endpoints with context length <= 8k
```

you would modify https://github.com/cline/cline/blob/main/src/api/providers/openrouter.ts

---

for example

https://www.reddit.com/r/SillyTavernAI/comments/1fi3baf/til_max_output_on_openrouter_is_actually_the/
> the "middle-out" method cuts out all the messages in the middle of your sent context so it fits into the supported one

https://github.com/moe-mizrak/laravel-openrouter
> `'transforms' => ['middle-out'],`

https://x.com/xanderatallah/status/1678506268657369090
> The OpenRouter Falcon API now allows "infinite-length" prompts using a "middle-out" strategy 😁
>
> Inspired by this paper, which showed that models with large context windows pay significantly less attention to the middle of prompts: [https://arxiv.org/abs/2307.03172](https://t.co/iD2PUp0olA)
>
> Since models are weak here anyway, we just cut out ChatML messages in the middle if input size > max input length :)

**Comment by lloydchang** [comment]
related https://github.com/cline/cline/discussions/743 https://github.com/cline/cline/issues/827 https://github.com/cline/cline/issues/849 https://github.com/cline/cline/pull/862 https://github.com/RooVetGit/Roo-Cline/pull/42

**Comment by lloydchang** [comment]
https://github.com/cline/cline/pull/862 fixes https://github.com/cline/cline/discussions/743
